feat: docs i18n support (continuation from #14818)

timhagn opened the pull request:
**Problem**

Our documentation is only written in English (and Chinese). This PR attempts to add support via Docusaurus' internationalization features.

Summary of Changes:

- [x] add i18n for the docs
- [x] merge in latest master
- [x] add Crowdin support & `crowdin:sync` command for CI
- [x] add languages setup in Crowdin & add German (de)
- [x] add failing locales in `docusaurus.config.js` as `// localesNotBuildingLocally: ["ko", "pt", "vi"],` comment
- [x] add `CROWDIN_PERSONAL_TOKEN` to Travis env variables
- [x] add call to `npm run crowdin:download` to Travis script

**Still TODO**

- [ ] look into failing locales
- [ ] add `CROWDIN_PERSONAL_TOKEN` to Vercel env variables
- [ ] Localize homepage (use json api)

jstarry commented:
Awesome @timhagn! Here's some high level feedback:

1. Keep translations out of the repo

Yes, I started this in https://github.com/solana-labs/solana/pull/17125 but I realized since then that it would be much better to download the translations as part of the doc CI build rather than include them in source. It only makes sense to keep them in source if we try to sync github changes back to crowdin, which docusaurus recommends not doing.

2. Add explicit headings

Since headings are translated, we need explicit headings to keep the slug the same and avoid breaking links. I don't think we need to include these explicit headings in source control so the best route is to run heading generation right before english source files are uploaded in CI. Info: https://docusaurus.io/docs/cli#docusaurus-write-heading-ids-sitedir

jstarry commented:
Versioning is hard so I think the best workflow for now is automatically uploading translations after each commit in the stable branch (currently v1.6) and setting up automatic PR's using Crowdin's github integration to sync translations back to the stable branch. This workflow means the translations will always be delayed and we won't have translations for beta and edge but I think that's ok for v1

timhagn commented:
Hi @jstarry,

1. removed the `i18n` folder from git's cache
2. added a `write-heading-ids` script for it

Pertaining the suggested process for versioning the docs, I just added `crowdin:sync` to the `before_install.sh` script at first,
but you're probably right, upload should only be done manually or for the `master` branch.
So I added a `crowdin:download` script & running only that in `before_install.sh`, or would you also want to remove that?

Best,

Tim.

jstarry commented on docs/package.json:
```diff
@@ -9,12 +9,20 @@
+    "crowdin": "crowdin",
+    "crowdin:download": "npm run write-i18n && npm run crowdin download && npm run format:fix",
+    "crowdin:sync": "npm run write-i18n && npm run crowdin upload && npm run crowdin download && npm run format:fix",
+    "prebuild": "npm run crowdin:download"
```
I would like to be able to build the english docs without downloading other translations

timhagn replied:
Removed it.

jstarry commented on docs/package.json:
```diff
@@ -9,12 +9,20 @@
     "lint": "set -ex; eslint .",
-    "lint:fix": "npm run lint -- --fix"
+    "lint:fix": "npm run lint -- --fix",
+    "crowdin": "crowdin",
```
Devs can use `npx crowdin ...` and the scripts can use `crowdin ...` directly, I don't think this shortcut is needed

timhagn replied:
You're probably right, will just add it to the docs.

timhagn replied:
Added a Translation paragraph to the README.

jstarry commented on docs/package.json:
```diff
@@ -9,12 +9,20 @@
-    "lint:fix": "npm run lint -- --fix"
+    "lint:fix": "npm run lint -- --fix",
+    "crowdin": "crowdin",
+    "crowdin:download": "npm run write-i18n && npm run crowdin download && npm run format:fix",
+    "crowdin:sync": "npm run write-i18n && npm run crowdin upload && npm run crowdin download && npm run format:fix",
```
I don't think we need to format files that aren't getting checked into git, and is there a need to write translations before downloading the latest?
```suggestion
    "crowdin:download": "crowdin download",
    "crowdin:upload": "npm run write-i18n && crowdin upload",
```

timhagn replied:
Linting broke the build in Travis when the files were not formatted correctly (while they were checked in).
Don't know if Travis will ignore them now when `i18n` is in `.gitignore`.

timhagn replied:
Added your changes, let's try it out in the next master build % ).

jstarry commented on docs/package.json:
```diff
@@ -9,12 +9,20 @@
+    "prebuild": "npm run crowdin:download"
   },
   "dependencies": {
+    "@crowdin/cli": "^3.5.2",
```
Do you mind bumping the below docusaurus versions to `2.0.0-beta.0`?

timhagn replied:
#done

jstarry commented on docs/sidebars.js:
```diff
@@ -45,7 +45,6 @@ module.exports = {
       "cli/deploy-a-program",
       "offline-signing",
       "offline-signing/durable-nonce",
-      "cli/usage",
```
why was this removed?

timhagn replied:
This must have fallen through the cracks while merging...

timhagn replied:
Re-added it.

jstarry reviewed (commented):
Thanks for the updates @timhagn! The upload should happen in the `build.sh` script before we generate the cli page and update release version numbers. You can look at the `publish-docs.sh` script for how to get the current channel. Uploading should only happen on the stable release* branch.

timhagn commented on docs/build.sh:
```diff
@@ -8,6 +8,12 @@ source ../ci/env.sh
 : "${rust_stable_docker_image:=}" # Pacify shellcheck
 
+# Synchronize translations with Crowdin
+if [[ -n $CI_TAG ]]; then
```
I did it according to:

```sh
if [[ -n $CI_TAG ]]; then
  PROJECT_NAME=docs-solana-com
else
...
```

from `publish-docs.sh`,  the switch for other tags afterwards tackles edge & beta.

But replaced it with:

```sh
# Fixated master branch as only one to download translations
ONLY_TRANSLATE_ON='master'
# Grep last name after last possible backslash for current branch
CURRENT_BRANCH=$(git symbolic-ref HEAD | sed -e 's,.*/\(.*\),\1,')

# Synchronize translations with Crowdin only on master branch
if [[ "$CURRENT_BRANCH" = "$ONLY_TRANSLATE_ON" ]]; then
```

timhagn replied:
Dang! Travis checks out commits not branches. Changed it to:

```sh
# Fixate master branch as only one to download translations & get its ref
ONLY_TRANSLATE_ON='master'
MASTER_REF="$(git rev-parse "$ONLY_TRANSLATE_ON")}"
# Get current ref
CURRENT_REF="$(cat ../.git/HEAD)"

# Synchronize translations with Crowdin only on master branch (ref)
if [[ "$CURRENT_REF" = "$MASTER_REF" ]]; then
  echo "Downloading & updating translations..."
  npm run crowdin:download
  npm run crowdin:upload
fi
```

but feeling like the original check might be less wonky % ).
Open for suggestions : )!

jstarry replied:
Hey, I mispoke earlier and we want to make sure that we are only uploading docs on the *stable* branch which you can use the `channel-info.sh` script to determine. I hope that simplifies the issue 😄

jstarry replied:
So the workflow I imagine is that every time we build the stable branch (in this case v1.6) the latest translations will get uploaded to crowdin. By the time we publish docs (usually on the next v1.6.x release) the script will download the latest translations which hopefully by then are up-to-date.

timhagn replied:
Just looked into it & it seems that on my system (Linux Mint 20.1, Kernel 5.4.0-73-generic) `semver.sh`
fails in both zsh & bash, my IDE / shellcheck complains about https://github.com/koalaman/shellcheck/wiki/SC2112 `'function' keyword is non-standard. Delete it.`.

So locally I'm currently unable to check for the stable branch , as the output of the 
`channel-info.sh` script is:

```sh
EDGE_CHANNEL=master
BETA_CHANNEL=v1.6
BETA_CHANNEL_LATEST_TAG=v1.6.9
STABLE_CHANNEL=v1.5
STABLE_CHANNEL_LATEST_TAG=v1.5.19
CHANNEL=
CHANNEL_LATEST_TAG=
```

(or what am I missing oO?) : (

t-nelson replied:
> (or what am I missing oO?) : (

STR?  I don't see `shellcheck` failing in any of the past Buildkite jobs

timhagn replied:
My IDE shellcheck fails with the aforementioned linked error message, not `shellcheck.sh` ; ).
But I still have no clue why `semver.sh` fails...
(By the way, what do you mean with "STR"? "Stands to Reason" is the only thing I could come up with...)

t-nelson replied:
Hmm, maybe just a difference in shellcheck version/config?  CI uses https://github.com/solana-labs/solana/blob/63b97729e6856be4f3955942ac8087c005ed202e/ci/shellcheck.sh

STR is "Steps to reproduce" :sweat_smile:

timhagn replied:
*lol* Sometimes not being a native English speaker makes it hard to understand some acronyms, thanks ^^.
Posted `semver.sh` on https://www.shellcheck.net/ (the site "koalaman" mentions in the [shellcheck repo](https://github.com/koalaman/shellcheck) & it gives me the same errors as my IDE (IntelliJ Studio), e.g.:

```sh
Line 3:
function semverParseInto() {
^-- SC2112: 'function' keyword is non-standard. Delete it.
```
But apart from that I'm still stuck at `channel-info.sh` not returning values for `CHANNEL` & `CHANNEL_LATEST_TAG`.
Should I just do a 

```sh
if [ "$CHANNEL" = "$STABLE_CHANNEL" ]; then
  echo "Downloading & updating translations..."
  npm run crowdin:download
  npm run crowdin:upload
fi

```

in the build script anyhow & be done with it % )? As it seems to work in CI I probably shouldn't mind my workstation - or should we rather look into it?

timhagn replied:
Hey @jstarry & @t-nelson,

any input on how to proceed?

Best,

Tim.

t-nelson replied:
I suspect the `shellcheck` issues you're seeing are from a lack of the `--shell=bash` arg we pass in CI.

`channel-info.sh` is _very_ presumptuous about being run under CI.  I have a branch with a WIP local CI wrapper.  It's not ready to merge though.  I'd recommend hacking values in around [here](https://github.com/solana-labs/solana/blob/b5302e76198b951e7d4f96fbf5f17e273f209dcd/ci/channel-info.sh#L113) and taking care not to merge them

jstarry replied:
all good here @timhagn?

timhagn replied:
Hi  @jstarry & @t-nelson,

wasn't able to do much yesterday, was already off when you replied (only ~2h sleep the night before -.-).

>  I'd recommend hacking values in around here and taking care not to merge them

Now I'm a little confused, which "values" should I hack into your linked `channel-info.sh`, @t-nelson  oO?
As that's identical to the one from `master` & this branch...

Best,

Tim.

t-nelson replied:
I'd recommend making the `build.sh` changes that you suggested in https://github.com/solana-labs/solana/pull/17190#discussion_r634740176, hack a  `CHANNEL` value in to `channel-info.sh` so you can check locally, pull the hack out once it's working as expected, then run it through CI

timhagn replied:
@t-nelson Replaced, adapted & tested it, should work : ).
Now we only have to look into getting the remaining languages working, but AFAIK @rmshea is at it.

jstarry replied:
Great @timhagn! I don't think we need to block on the remaining locales. Any other blockers for this going in?

timhagn replied:
@jstarry Nope, should be good to go, LGTM : )!

t-nelson commented on docs/build.sh:
```diff
@@ -8,14 +8,27 @@ source ../ci/env.sh
 
+# Fixate master branch as only one to download translations & get its ref
+ONLY_TRANSLATE_ON='master'
+MASTER_REF="$(git rev-parse "$ONLY_TRANSLATE_ON")}"
```
Stray `}` ?

timhagn replied:
Thanks! Fixed.

t-nelson commented on docs/build.sh:
```diff
@@ -8,14 +8,27 @@ source ../ci/env.sh
 
 # Build from /src into /build
-npm run build
+#npm run build
```
Lets kill the block instead of committing commented code

timhagn replied:
Ups, that was a remainder from testing it % ). Uncommented again.

t-nelson reviewed (dismissed):
`*.sh` changes lgtm!

jstarry reviewed (changes requested):
Please remove the zip file and revert the heading id change

timhagn commented:
> Please remove the zip file and revert the heading id change

Removed the zip file - no clue where that broken one came from, `git blame` didn't show me.
Which ID do you mean?

jstarry commented:
> > Please remove the zip file and revert the heading id change
> 
> Removed the zip file - no clue where that broken one came from, `git blame` didn't show me.
> Which ID do you mean?

Great! These are the id's I'm referring to: 

<img width="572" alt="Screen Shot 2021-05-25 at 2 37 24 PM" src="https://user-images.githubusercontent.com/1076145/119571967-df8ff700-bd66-11eb-9037-b2b55e616ebe.png">

Instead of checking in these id's, they should be generated right before crowdin upload so that us devs don't have to worry about them.

timhagn commented:
Thnx @jstarry!

Thought you meant an HTML header ^^. Used this regex to find and replace them: `\ \{\#.*`, don't think it broke anything (I hope so, indeed ; ). `crowdin:upload` autogenerates them on build.

Best,

Tim.

jstarry reviewed (approved):
looks great, sorry for the approval delay! Let's give this a shot 🚢

timhagn commented:
Great, LGTM : )!
